src/ide/tgui.ts

```diff
@@ -1146,14 +1167,401 @@ export let tgui = (function() {
 						return false;   // important: consume clicks
 					},
 			});
+
+	// See `createModal` for a detailed description of the objects
+	// that are stored here
 	let modal = new Array();
-	
+
+	function roundToPhysicalPixel(virtual_px: number)
+	{
+		return (virtual_px*window.devicePixelRatio | 0)/window.devicePixelRatio;
+	}
+
+	// Center a dialog at the center of the window
+	// Dialogs are aligned at integral pixel coordinates to
+	// avoid pixel-jittering
+	function centerModalDialog(dlg)
+	{
+		let dlg_width, dlg_height;
+		let scr_width = window.innerWidth, scr_height = window.innerHeight;
+		if(dlg.hasOwnProperty("minsize") && dlg.hasOwnProperty("scalesize"))
+		{
+			dlg_width = roundToPhysicalPixel(Math.min(Math.max(dlg.minsize[0], dlg.scalesize[0]*scr_width), scr_width));
+			dlg_height = roundToPhysicalPixel(Math.min(Math.max(dlg.minsize[1], dlg.scalesize[1]*scr_height), scr_height));
+			
+			dlg.dom.style["width"] = dlg_width+"px";
+			dlg.dom.style["height"] = dlg_height+"px";
+		}
+		else
+		{
+			let rect = dlg.dom.getBoundingClientRect();
+			dlg_width = rect.width;
+			dlg_height = rect.height;
+		}
+		dlg.dom.style["left"] = roundToPhysicalPixel((scr_width-dlg_width)/2)+"px";
+		dlg.dom.style["top"]  = roundToPhysicalPixel((scr_height-dlg_height)/2)+"px";
+	}
+
+	// Center all dialogs whenever the window changed
+	function centerAllModalDialogs()
+	{
+		for(let i = 0; i < modal.length; ++i) centerModalDialog(modal[i]);
+	}
+
+	// TODO: merge this with arrangePanels, that is also installed as a "resize"-callback
+	window.addEventListener("resize", centerAllModalDialogs);
+
+
+	// Create a modal dialog. Similar to createPanel
+	// The description object has the following fields:
+	// - title:             text in the title bar
+	// - scalesize:         [width, heigth] scaled size of the dialog
+	// - minsize:           [width, height] minimum size of the dialog, 
+	//                      when the whole viewport is smaller, the viewport size is used
+	// - contentstyle:      object to add/override some styles to/of the content element
+	// - buttons:           list of [name, text] pairs like [["delete", "Delete File"], ["cancel", "Cancel"]],
+	//                      if this is not given, there is no button bar
+	//                      at the bottom. The first element, the name, is passed to onButton and
+	//                      might be referred to in the defaultButton attribute.
+	//                      The second is only used as the text displayed on the button.
+	//                      It might be a translated text.
+	//                      Note that an object/a dictionary is not used, because the order of the elements
+	//                      is important.
```

Comment:
I do not think that this is a good idea, because the text on the button might be generated dynamically and the text itself might be a translation into another language like German. When doing this, the code calling `tgui.createModal`/`tgui.msgBox` is going to get more complex and the text in the `buttons` and in `onButton` might get out of sync.

Reply:
I see your point, it's probably just that I don't like using tuples for this. I guess it's the right choice then for simplicity 👍.

Reply:
I thought about it and came to the conclusion, that objects instead of tuples are more expressive and would resolve the problem of the button identifier strings, because every button could have its own handler now.